x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/expression/Expressions.java

```diff
@@ -132,8 +132,16 @@ public static String name(Expression e) {
         return e instanceof NamedExpression ne ? ne.name() : e.sourceText();
     }
 
-    public static boolean isNull(Expression e) {
-        return e.dataType() == DataType.NULL || (e.foldable() && e.fold() == null);
+    /**
+     * Is this {@linkplain Expression} <strong>guaranteed</strong> to have
+     * only the {@code null} value. {@linkplain Expression}s that
+     * {@link Expression#fold()} to {@code null} <strong>may</strong>
+     * return {@code false} here, but should <strong>eventually</strong> be folded
+     * into a {@link Literal} containing {@code null} which will return
+     * {@code true} from here.
+     */
+    public static boolean isGuaranteedNull(Expression e) {
```

Comment:
Nit: `isGuaranteedNull` makes me wonder what is "null" and more relevant when is it not guaranteed :) 
I'd find something like `isNullTypeOrLiteral` clearer, but just a nitty nit.

Reply:
Fair enough. I think I'm going to keep it because everything feels not so great. Maybe `valueIsGuaranteedNull` or something? I'm not sure that makes it more descriptive.